formats/index/decoder.go

```diff
@@ -0,0 +1,389 @@
+	version, err := validateHeader(d.r)
+	if err != nil {
+		return err
+	}
+
+	idx.Version = version
+
+	if err := binary.Read(d.r, binary.BigEndian, &idx.EntryCount); err != nil {
+		return err
+	}
+
+	if err := d.readEntries(idx); err != nil {
+		return err
+	}
+
+	return d.readExtensions(idx)
+}
+
+func (d *Decoder) readEntries(idx *Index) error {
+	for i := 0; i < int(idx.EntryCount); i++ {
+		e, err := d.readEntry(idx)
+		if err != nil {
+			return err
+		}
+
+		d.lastEntry = e
+		idx.Entries = append(idx.Entries, *e)
+	}
+
+	return nil
+}
+
+func (d *Decoder) readEntry(idx *Index) (*Entry, error) {
+	e := &Entry{}
+
+	var msec, mnsec, sec, nsec uint32
+
+	flow := []interface{}{
+		&msec, &mnsec,
+		&sec, &nsec,
+		&e.Dev,
+		&e.Inode,
+		&e.Mode,
+		&e.UID,
+		&e.GID,
+		&e.Size,
+		&e.Hash,
+		&e.Flags,
+	}
+
+	if err := readBinary(d.r, flow...); err != nil {
+		return nil, err
+	}
+
+	read := 62
+	e.CreatedAt = time.Unix(int64(msec), int64(mnsec))
+	e.ModifiedAt = time.Unix(int64(sec), int64(nsec))
+	e.Stage = Stage(e.Flags>>12) & 0x3
+
+	if e.Flags&EntryExtended != 0 {
+		read += 2
+		var extended uint16
+		if err := readBinary(d.r, &extended); err != nil {
+			return nil, err
+		}
+
+		e.IntentToAdd = extended&intentToAddMask != 0
+		e.SkipWorktree = extended&skipWorkTreeMask != 0
+	}
+
+	if err := d.readEntryName(idx, e); err != nil {
+		return nil, err
+
+	}
+
+	return e, d.padEntry(idx, e, read)
+}
+
+func (d *Decoder) readEntryName(idx *Index, e *Entry) error {
+	var name string
+	var err error
+
+	switch idx.Version {
+	case 2, 3:
+		name, err = d.doReadEntryName(e)
+	case 4:
+		name, err = d.doReadEntryNameV4()
+	}
+
+	if err != nil {
+		return err
+	}
+
+	e.Name = name
+	return nil
+}
+
+func (d *Decoder) doReadEntryNameV4() (string, error) {
+	l, err := readVariableWidthInt(d.r)
+	if err != nil {
+		return "", err
+	}
+
+	var base string
+	if d.lastEntry != nil {
+		base = d.lastEntry.Name[:len(d.lastEntry.Name)-int(l)]
+	}
+
+	name, err := readUntil(d.r, 0)
+	if err != nil {
+		return "", err
+	}
+
+	return base + string(name), nil
+}
+
+func (d *Decoder) doReadEntryName(e *Entry) (string, error) {
+	pLen := e.Flags & nameMask
+
+	name := make([]byte, int64(pLen))
+	if err := binary.Read(d.r, binary.BigEndian, &name); err != nil {
+		return "", err
+	}
+
+	return string(name), nil
+}
+
+// Index entries are padded out to the next 8 byte alignment
+// for historical reasons related to how C Git read the files.
+func (d *Decoder) padEntry(idx *Index, e *Entry, read int) error {
+	if idx.Version == 4 {
+		return nil
+	}
+
+	entrySize := read + len(e.Name)
+	padLen := 8 - entrySize%8
+	if _, err := io.CopyN(ioutil.Discard, d.r, int64(padLen)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *Decoder) readExtensions(idx *Index) error {
+	var err error
+	for {
+		err = d.readExtension(idx)
+		if err != nil {
+			break
+		}
+	}
+
+	if err == io.EOF {
+		return nil
+	}
+
+	return nil
+}
```

Comment:
The are some checks missing here, otherwise the last 5 lines can be replaced by:

```
return nil
```

And whole method does not need to return anything.